posts/2023-09-30-keycloak-with-openliberty.adoc

```diff
@@ -0,0 +1,126 @@
+seo-description: Open Liberty can easily integrate with Keycloak, which can be used as an OpenID Connect Provider, using Jakarta Security and MicroProfile JWT.
+blog_description: "Open Liberty can easily integrate with Keycloak, which can be used as an OpenID Connect Provider, using Jakarta Security and MicroProfile JWT."
+open-graph-image: https://openliberty.io/img/twitter_card.jpg
+---
+= Protect your applications with Jakarta Security, MicroProfile JWT, and Keycloak
+Jimmy Wu <https://github.com/jimmy1wu>
+:imagesdir: /
+:url-prefix:
+:url-about: /
+//Blank line here is necessary before starting the body of the post.
+
+
+Open Liberty can easily integrate with Keycloak, which can be used as an OpenID Connect Provider, using Jakarta Security and MicroProfile JWT.
+In this blog post, a Keycloak server has been setup with a realm called `openliberty` which contains a client called `sample-openliberty-keycloak` and the realm roles of `admin` and `user`. Additionally in the `sample-openliberty-keycloak` client, client authentication has been enabled, `http://localhost:9090/Callback` has been added as a valid redirect URI, and the `microprofile-jwt` client scope has been set to `Default`.
+
+== Obtaining an access token from Keycloak using Jakarta Security
+
+With the new `@OpenIdAuthenticationMechanismDefinition` annotation introduced in link:https://openliberty.io/docs/latest/reference/feature/appSecurity-5.0.html[Jakarta Security 3.0], you can easily authenticate users of your application with Keycloak and obtain an access token.
+This example shows how the `@OpenIdAuthenticationMechanismDefinition` can be configured to setup an authentication flow with Keycloak.
+
+[source,java]
+----
+@OpenIdAuthenticationMechanismDefinition(
+        providerURI = "http://localhost:8080/realms/openliberty/.well-known/openid-configuration",
+        clientId = "sample-openliberty-keycloak",
+        clientSecret = "x4fRVAhk49TKDqVlzIt4q9oh8DSWfePt",
+        redirectToOriginalResource = true,
+        logout = @LogoutDefinition(notifyProvider = true))
+----
+
+In this example, the `providerURI` is the discovery endpoint for the `openliberty` realm, the `clientId` is the client ID of the `sample-openliberty-keycloak` client, and the `clientSecret` is the secret belonging to the `sample-openliberty-keycloak` client. By default, the redirect URI is set to `http://localhost:9090/Callback` and `redirectToOriginalResource` is set to `true` to redirect users from the redirect URI back to the originally requested resource. Lastly, `notifyProvider` in the `@LogoutDefinition` is set to `true` to also log the user out of Keycloak when a logout occurs in your Open Liberty application.
+
+Now, with this annotation set up, your REST endpoints can be protected using the `@RolesAllowed` annotation which will trigger the authentication flow when a user tries to access the endpoint.
```

Comment:
More why here would be nice - why is protection needed? what do roles help us with and how?